### ISSUE TITLE
Stuck in a single point

### ISSUE THREAD

**Issue by RenKelvin** [TomTheBotter/Pokemon-Go-Bot-Working-Hack-API#60]
After running some time, the avatar will stuck in a single point. Is this a bug?

**Comment by Sytharix** [comment]
You shouldn't have your game open on your phone and this script running at the same time (Assuming this is what you're doing). This will raise red flags with Niantic and likely get your account soft banned. Try interacting with pokestops around you, if you can't then your account's been soft banned.

**Comment by 1azlef1** [comment]
Hey I have the same problem and no I wasn't running the game at the same time on my phone and my account still works normally the bot just gets stuck on a certain location.

**Comment by Sytharix** [comment]
@1azlef1 Have you tried restarting the script? Make sure if you're not running it for the first time you use `python pokebot.py -i 0 --cache` or `python pokebot.py -i 1 --cache` (Depending on if you're using Google or not).

**Comment by 1azlef1** [comment]
Yes I have restarted the script at least five times and everytime it goes to "

> Walking to fort at 40.768795,-73.977854

" it stops and repeats "

> Walking to fort at 40.768795,-73.977854

". It still catches pokémons from time to time but stays at that point. Also I have used --cache the first  time and didn't do it the next time.

**Comment by Daguerro** [comment]
I have the same problem, no matter, if I run it with or without --cache. After 5-10 minutes, it will stop walking around and stays at the same location. A restart fix this for another 5-10 minutes. Same for google and PTC accounts and starting location doesn't matter either...

**Comment by dubcanada** [comment]
Also staying in the same location for me. It seems to run spin_near_fort and if there is no fort nearby it does nothing. It should just randomly walk around and if there happens to be a fort in it's area then do something.

**Comment by DE816** [comment]
I also confirm this problem.

**Comment by Jay1** [comment]
Same problem here, not much info to add yet, I'm trying to investigate. Restarting fix it everytime so far so I expect this is not PoGo related

**Comment by aeeh** [comment]
Have the same issues.
How to solve.

**Comment by DE816** [comment]
> Same problem here, not much info to add yet, I'm trying to investigate. Restarting fix it everytime so far so I expect this is not PoGo related

Edit: I'm getting less and less pokemon until everything stop, so I assumed it was related to step size. So far a step of 2 can't last more than 20 seconds. Step of 50 gave me 2-3 minutes at best.

My step size is at 100 and I am no longer having issues atm.

**Comment by TomTheBotter** [comment]
@DE816 Thanks for that...I didn't realize that as my bot previously was using a step size of 200 and that's how I was able to level up pretty fast. I'll look into it more and send out a working solution

**Comment by aeeh** [comment]
Using 200, he stucks at one point for 3-4 minutes and after some time he's going forward to the next point.
Strange behaviour.

**Comment by Jay1** [comment]
Well step of 70 gave me 20 minutes, longest I had yet
so far best results are 70>50>100>200>2>1

I can't find any correlation yetr

**Comment by aeeh** [comment]
70 isn't working well for me.
Hmm, dunno to solve this. :/

**Comment by 1azlef1** [comment]
I don't know if this matters but the bot always gets stuck on the same coordinates. I set my location as central park.

**Comment by DE816** [comment]
@TomTheBotter Is 1 extreamly slow and 200 very fast? I am just wondering how steps work for this bot.

**Comment by Jay1** [comment]
@DE816  1 walking speed ish as I can tell from pokevision,  200 is not very fast more like bicycle

**Comment by Maximum-throwaway** [comment]
Also getting this issue. After awhile the bot just gets stuck with dozens of pokestops and pokemon around.

**Comment by Daguerro** [comment]
this release fixed it [https://github.com/etse/Pokemon-Go-Bot-Working-Hack-API](url)

**Comment by aeeh** [comment]
@Daguerro the link doesn't work :/

**Comment by NowNewStart** [comment]
[Working Link](https://github.com/etse/Pokemon-Go-Bot-Working-Hack-API)

**Comment by charredchar** [comment]
The one from etse link gives me an error.

2016-07-24 16:46:45,630 [   pokebot] [ERROR] Main loop has an ERROR, restarting unsupported operand type(s) for %: 'int' and 'NoneType'

The version from TomTheBotter still gets stuck at stops, it seems to be the same stops every time too even after a restart.

**Comment by NowNewStart** [comment]
@charredchar This happens when you dont have any pokeballs

**Comment by charredchar** [comment]
@NowNewStart Except I do.

**Comment by xthedeus** [comment]
@NowNewStart I'm getting the same error. Not getting it without those new modifications.

**Comment by zyklus** [comment]
I'm fairly certain @NowNewStart's patch doesn't address this problem, it simply avoids it in most cases.

The issue exists in the game client as well -- there are simply some "unspinnable" pokestops, for whatever reason.  The bot seems to be getting stuck on these.

The solution is to not take the server response as 100% authoritative.  The bot needs to keep track of the last pokestop it spun, and if the server responds with that one as spinnable on the next response then flag it as "bad", filter it from the results and move on.

I'd do it, but I don't know python and someone else could probably write it much faster :)

**Comment by xthedeus** [comment]
@zyklus Maybe if you combine the fort_id with a timestamp in a map and keep track of the last 10 forts visited or so. Then make a lookup in the map and check if the current fort has been visited within the last x minutes (can't remember how long time it takes for a fort to be spinnable).

It has been too long time since I wrote some python but I'm very interested in the discussion to find a proper approach! :)

**Comment by zyklus** [comment]
@xthedeus - In my experience with the app itself, some pokestops are just bad (for instance there is one near my house that I've never been able to spin).  They'll probably be fixed at some point, but I think a blacklist for the purpose of this bot would work.  Otherwise the bot would just go back to that one and try it again after whatever timeout.

**Comment by xthedeus** [comment]
Ah alright! I've never seen that but you might be right. A blacklist would also be possible. Then the approach would be to just store the id of the fort and check that a fort has not been blacklisted before "walking". Maybe storing the blacklisted forts in a file would make sense?

**Comment by zyklus** [comment]
So thinking about this a bit more, with the reliability of Niantic's servers I'd be slightly concerned about blacklisting forts that the server simply choked on for whatever reason.

Perhaps:

```json
blacklist : [{
    "id" : 123,
    "failed_attempts" : 2,
    "last_attempt" : "...date..."
}]
```

I think some simple rules on this would work:
- don't try again for an increasing time span (e.g. `5 ^ failed_attempts` minutes)
- clear if it ever succeeds

Then how do we add something to the blacklist?

```js
lastSpun = fort_id

// next tick
if(isGreen(lastSpun)) {
  markAsBad(lastSpun)
}
```
